File: cpp/src/prefix_beam_search.cpp
#include "prefix_beam_search.h"

#include <algorithm>
#include <cmath>
#include <iterator>
#include <limits>
#include <map>
#include <unordered_set>

namespace asr_decoder::internal {
namespace {

std::vector<std::pair<int, double>> top_k(const float* row, size_t size,
                                          size_t count) {
  std::vector<std::pair<int, double>> result;
  result.reserve(size);
  for (size_t token = 0; token < size; ++token) {
    result.emplace_back(static_cast<int>(token), row[token]);
  }
  const size_t kept = std::min(count, result.size());
  std::partial_sort(result.begin(), result.begin() + kept, result.end(),
                    [](const auto& left, const auto& right) {
                      if (left.second != right.second) {
                        return left.second > right.second;
                      }
                      return left.first > right.first;
                    });
  result.resize(kept);
  return result;
}

bool prepare_search(DecoderState& state, const float* values, size_t frames,
                    size_t vocabulary_size, DecodeOptions options,
                    SearchSettings& settings) {
  state.error.clear();
  if ((!values && frames) || vocabulary_size == 0 ||
      static_cast<size_t>(state.config.blank_id) >= vocabulary_size ||
      (!state.graph.empty() &&
       static_cast<size_t>(state.graph.maximum_token()) >= vocabulary_size)) {
    state.error = "invalid log-probability matrix or token ID";
    return false;
  }
  if (state.vocabulary_size && state.vocabulary_size != vocabulary_size) {
    state.error = "vocabulary_size changed during a stream";
    return false;
  }
  settings = settings_for(state.config.decoding_quality);
  if (options.beam_size) settings.beam = options.beam_size;
  if (options.token_beam_size) settings.token_beam = options.token_beam_size;
  settings.token_beam = std::min(settings.token_beam, vocabulary_size);
  if (settings.beam == 0 || settings.token_beam == 0) {
    state.error = "beam sizes must be positive";
    return false;
  }
  if (state.beam_size && state.beam_size != settings.beam) {
    state.error = "beam_size changed during a stream";
    return false;
  }
  if (state.token_beam_size && state.token_beam_size != settings.token_beam) {
    state.error = "token_beam_size changed during a stream";
    return false;
  }
  for (size_t index = 0; index < frames * vocabulary_size; ++index) {
    if (std::isnan(values[index]) ||
        values[index] == std::numeric_limits<float>::infinity()) {
      state.error =
          "log probabilities must not contain NaN or positive infinity";
      return false;
    }
  }
  for (size_t frame = 0; frame < frames; ++frame) {
    bool has_finite_value = false;
    for (size_t token = 0; token < vocabulary_size; ++token) {
      has_finite_value |=
          std::isfinite(values[frame * vocabulary_size + token]);
    }
    if (!has_finite_value) {
      state.error = "every frame must contain a finite log probability";
      return false;
    }
  }
  state.vocabulary_size = vocabulary_size;
  state.beam_size = settings.beam;
  state.token_beam_size = settings.token_beam;
  return true;
}

void advance_frame(DecoderState& state, const float* row,
                   size_t vocabulary_size, const SearchSettings& settings) {
  const ContextGraph* graph = state.graph.empty() ? nullptr : &state.graph;
  auto acoustic = top_k(row, vocabulary_size, settings.token_beam);
  if (acoustic.empty()) return;
  const double best = acoustic.front().second;
  const double blank = row[state.config.blank_id];
  const double threshold =
      state.policy.context_prune_threshold * gating_factor(acoustic, blank);

  std::map<int, double> candidates;
  for (const auto& [token, probability] : acoustic) {
    if (probability >= best - settings.prune_threshold ||
        token == state.config.blank_id) {
      candidates[token] = probability;
    }
  }
  candidates[state.config.blank_id] = blank;
  if (graph) {
    std::unordered_set<int> injected;
    for (const auto& hypothesis : state.hypotheses) {
      for (int token :
           graph->candidates(hypothesis.second.context_state,
                             state.policy.max_injected_candidates)) {
        if (injected.size() == state.policy.max_injected_candidates) break;
        injected.insert(token);
      }
    }
    for (int token : injected) {
      const double probability = row[token];
      if (probability >= best - threshold) candidates[token] = probability;
    }
  }

  std::map<std::vector<int>, PrefixScore> next;
  ++state.processed_frames;
  for (const auto& [token, probability] : candidates) {
    for (const auto& [prefix, source] : state.hypotheses) {
      if (token == state.config.blank_id) {
        PrefixScore& destination = next[prefix];
        destination.blank =
            log_add(destination.blank, source.acoustic() + probability);
        const double viterbi = source.viterbi() + probability;
        if (viterbi > destination.viterbi_blank) {
          destination.viterbi_blank = viterbi;
          destination.blank_alignment = source.alignment();
        }
        copy_context(source, destination, graph != nullptr);
        continue;
      }

      if (!prefix.empty() && prefix.back() == token) {
        PrefixScore& repeated = next[prefix];
        repeated.nonblank =
            log_add(repeated.nonblank, source.nonblank + probability);
        if (source.viterbi_nonblank + probability > repeated.viterbi_nonblank) {
          repeated.viterbi_nonblank = source.viterbi_nonblank + probability;
          repeated.nonblank_alignment = source.nonblank_alignment;
          if (!repeated.nonblank_alignment.empty()) {
            TokenSpan& span = repeated.nonblank_alignment.back();
            span.end = state.processed_frames;
            span.log_probability = std::max(span.log_probability, probability);
          }
        }
        copy_context(source, repeated, graph != nullptr);

        std::vector<int> extended_prefix = prefix;
        extended_prefix.push_back(token);
        PrefixScore& after_blank = next[extended_prefix];
        after_blank.nonblank =
            log_add(after_blank.nonblank, source.blank + probability);
        if (source.viterbi_blank + probability > after_blank.viterbi_nonblank) {
          after_blank.viterbi_nonblank = source.viterbi_blank + probability;
          after_blank.nonblank_alignment = source.blank_alignment;
          after_blank.nonblank_alignment.push_back(
              {token, state.processed_frames - 1, state.processed_frames,
               probability});
        }
        advance_context(graph, state.policy, source, after_blank, token);
        continue;
      }

      std::vector<int> extended_prefix = prefix;
      extended_prefix.push_back(token);
      PrefixScore& destination = next[extended_prefix];
      destination.nonblank =
          log_add(destination.nonblank, source.acoustic() + probability);
      if (source.viterbi() + probability > destination.viterbi_nonblank) {
        destination.viterbi_nonblank = source.viterbi() + probability;
        destination.nonblank_alignment = source.alignment();
        destination.nonblank_alignment.push_back(
            {token, state.processed_frames - 1, state.processed_frames,
             probability});
      }
      advance_context(graph, state.policy, source, destination, token);
    }
  }
  state.hypotheses.assign(std::make_move_iterator(next.begin()),
                          std::make_move_iterator(next.end()));
  std::sort(state.hypotheses.begin(), state.hypotheses.end(),
            [](const auto& left, const auto& right) {
              if (left.second.total() != right.second.total()) {
                return left.second.total() > right.second.total();
              }
              return left.first < right.first;
            });
  if (state.hypotheses.size() > settings.beam) {
    state.hypotheses.resize(settings.beam);
  }
}

DecodeResult build_result(const DecoderState& state, DecodeOptions options) {
  DecodeResult result;
  for (const auto& hypothesis : state.hypotheses) {
    const PrefixScore& score = hypothesis.second;
    const auto& alignment = score.alignment();
    std::vector<int> tokens;
    std::vector<Timestamp> timestamps;
    std::vector<double> probabilities;
    for (const TokenSpan& span : alignment) {
      tokens.push_back(span.token);
      Timestamp timestamp{span.start, span.end};
      if (state.config.frame_shift_ms > 0.0) {
        timestamp.start_ms = span.start * state.config.frame_shift_ms;
        timestamp.end_ms = span.end * state.config.frame_shift_ms;
      }
      timestamps.push_back(timestamp);
      if (options.return_token_probabilities) {
        probabilities.push_back(std::exp(span.log_probability));
      }
    }
    result.tokens.push_back(std::move(tokens));
    result.timestamps.push_back(std::move(timestamps));
    if (options.return_token_probabilities) {
      result.probabilities.push_back(std::move(probabilities));
    }
    result.scores.push_back({score.acoustic(), score.context(), score.total()});
  }
  return result;
}

}  // namespace

DecodeResult run_prefix_beam_search(DecoderState& state, const float* values,
                                    size_t frames, size_t vocabulary_size,
                                    DecodeOptions options) {
  DecodeResult result;
  if (!state.config_error.empty()) return result;
  SearchSettings settings;
  if (!prepare_search(state, values, frames, vocabulary_size, options,
                      settings)) {
    return result;
  }
  for (size_t frame = 0; frame < frames; ++frame) {
    advance_frame(state, values + frame * vocabulary_size, vocabulary_size,
                  settings);
  }
  result = build_result(state, options);
  if (options.finalize) state.reset();
  return result;
}

}  // namespace asr_decoder::internal

File: rust/src/prefix_beam_search.rs
use crate::context_policy::AdaptiveContextGate;
use crate::decoder::{DecodeMode, SearchConfig};
use crate::prefix_score::materialize_alignment;
use crate::{
    CTCDecoder, DecodeResult, DecoderError, GatingDiagnostics, HypothesisScore, LogProbabilities,
    PrefixBeamSearchOptions,
};

mod candidates;
mod transitions;

use candidates::{top_k, PreparedFrame};

impl CTCDecoder {
    pub fn prefix_beam_search(
        &mut self,
        input: LogProbabilities<'_>,
        options: PrefixBeamSearchOptions,
    ) -> Result<DecodeResult, DecoderError> {
        self.validate_input(input)?;
        let search_config = self.resolve_search_config(
            options.beam_size,
            options.token_beam_size,
            input.vocabulary_size(),
        )?;
        let frames = self.prepare_frames(input, search_config);
        self.lock_decode_mode(DecodeMode::PrefixBeam)?;
        if self.search_config.is_none() {
            self.search_config = Some(search_config);
        }

        for (frame_offset, frame) in frames.into_iter().enumerate() {
            self.processed_frames += 1;
            let (acoustic_scale, confidence_factor) = self.frame_gating(&frame.gating, frame.blank);
            let mut context_candidates = self.collect_context_candidates();
            let active_factor = AdaptiveContextGate::new(&self.context_policy.gating)
                .active_context_factor(
                    context_candidates.competing_count,
                    self.context_policy.max_injected_candidates,
                );
            let gating_factor = acoustic_scale * confidence_factor * active_factor;
            if !frame.gating.is_empty() && frame.blank < frame.values[0] {
                self.diagnostics.gating_factor_sum += gating_factor;
                self.diagnostics.nonblank_gating_frames += 1;
            }
            let context_threshold =
                self.context_policy.context_token_prune_threshold * gating_factor;
            let candidates = self.frame_candidates(
                input.frame(frame_offset),
                &frame,
                search_config,
                &mut context_candidates,
                context_threshold,
            );
            self.advance_prefix_frame(&candidates, search_config.beam_size);
        }

        let result = self.prefix_response(search_config, options);
        if options.finalize {
            self.reset();
        }
        Ok(result)
    }

    fn resolve_search_config(
        &self,
        beam_size: Option<usize>,
        token_beam_size: Option<usize>,
        vocabulary_size: usize,
    ) -> Result<SearchConfig, DecoderError> {
        if let Some(current) = self.search_config {
            if current.vocabulary_size != vocabulary_size {
                return Err(DecoderError::SearchConfigChanged("vocabulary_size"));
            }
            if beam_size.is_some_and(|value| value != current.beam_size) {
                return Err(DecoderError::SearchConfigChanged("beam_size"));
            }
            let requested = token_beam_size.map(|value| value.min(vocabulary_size));
            if requested.is_some_and(|value| value != current.token_beam_size) {
                return Err(DecoderError::SearchConfigChanged("token_beam_size"));
            }
            return Ok(current);
        }
        let (default_beam, default_token_beam, threshold) = self.decoding_quality.search_settings();
        let beam_size = beam_size.unwrap_or(default_beam);
        let token_beam_size = token_beam_size
            .unwrap_or(default_token_beam)
            .min(vocabulary_size);
        if beam_size == 0 {
            return Err(DecoderError::InvalidConfig("beam_size must be at least 1"));
        }
        if token_beam_size == 0 {
            return Err(DecoderError::InvalidConfig(
                "token_beam_size must be at least 1",
            ));
        }
        Ok(SearchConfig {
            beam_size,
            token_beam_size,
            token_prune_threshold: threshold,
            vocabulary_size,
        })
    }

    fn prepare_frames(
        &self,
        input: LogProbabilities<'_>,
        config: SearchConfig,
    ) -> Vec<PreparedFrame> {
        (0..input.frames())
            .map(|frame_index| {
                let row = input.frame(frame_index);
                let top = top_k(row, config.token_beam_size);
                let (indices, values): (Vec<_>, Vec<_>) = top.into_iter().unzip();
                let gating = if self.context_graph.is_some() && self.context_policy.gating.enabled {
                    top_k(row, row.len().min(8))
                        .into_iter()
                        .map(|(_, value)| value)
                        .collect()
                } else {
                    Vec::new()
                };
                PreparedFrame {
                    values,
                    indices,
                    blank: f64::from(row[self.blank_id]),
                    gating,
                }
            })
            .collect()
    }

    fn frame_gating(&mut self, top: &[f64], blank: f64) -> (f64, f64) {
        if top.is_empty() {
            return (1.0, 1.0);
        }
        let stats = AdaptiveContextGate::new(&self.context_policy.gating).analyze_frame(top, blank);
        self.diagnostics.gating_acoustic_scale_sum += stats.0;
        self.diagnostics.gating_frames += 1;
        stats
    }

    fn prefix_response(
        &self,
        search_config: SearchConfig,
        options: PrefixBeamSearchOptions,
    ) -> DecodeResult {
        let mut hypotheses = self.hypotheses.clone();
        hypotheses.sort_by(|left, right| right.1.total_score().total_cmp(&left.1.total_score()));
        let alignments: Vec<_> = hypotheses
            .iter()
            .map(|(_, score)| materialize_alignment(score.alignment()))
            .collect();
        DecodeResult {
            tokens: alignments
                .iter()
                .map(|alignment| alignment.0.clone())
                .collect(),
            timestamps: alignments
                .iter()
                .map(|alignment| self.timestamps(&alignment.1))
                .collect(),
            probabilities: options.return_token_probabilities.then(|| {
                alignments
                    .iter()
                    .map(|alignment| alignment.2.iter().map(|score| score.exp()).collect())
                    .collect()
            }),
            scores: options.return_scores.then(|| {
                hypotheses
                    .iter()
                    .map(|(_, score)| HypothesisScore {
                        acoustic: score.acoustic_score(),
                        contextual_bias: score.contextual_score(),
                        total: score.total_score(),
                    })
                    .collect()
            }),
            gating: options
                .return_gating_diagnostics
                .then(|| self.gating_diagnostics(search_config)),
        }
    }

    fn gating_diagnostics(&self, config: SearchConfig) -> GatingDiagnostics {
        let enabled = self.context_graph.is_some() && self.context_policy.gating.enabled;
        GatingDiagnostics {
            contextual_biasing: self.context_graph.is_some(),
            adaptive_context_gating: enabled,
            gating_frames: if enabled {
                self.diagnostics.gating_frames
            } else {
                0
            },
            mean_gating_acoustic_scale: (enabled && self.diagnostics.gating_frames > 0).then(
                || {
                    self.diagnostics.gating_acoustic_scale_sum
                        / self.diagnostics.gating_frames as f64
                },
            ),
            mean_gating_factor: (enabled && self.diagnostics.nonblank_gating_frames > 0).then(
                || {
                    self.diagnostics.gating_factor_sum
                        / self.diagnostics.nonblank_gating_frames as f64
                },
            ),
            nonblank_gating_frames: if enabled {
                self.diagnostics.nonblank_gating_frames
            } else {
                0
            },
            evaluated_context_candidates: self.diagnostics.evaluated_context_candidates,
            gathered_context_candidates: self.diagnostics.gathered_context_candidates,
            max_context_candidates_per_frame: self.diagnostics.max_context_candidates_per_frame,
            context_candidate_limit_hits: self.diagnostics.context_candidate_limit_hits,
            word_boundary_aware_context: !self.word_boundary_token_ids.is_empty(),
            boundary_suppressed_context_candidates: self
                .diagnostics
                .boundary_suppressed_context_candidates,
            beam_size: config.beam_size,
            token_beam_size: config.token_beam_size,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{ContextPolicy, DecoderConfig};

    #[test]
    fn streaming_matches_one_shot() {
        let values = [-5.0, 0.0, -5.0, 0.0, -5.0, -5.0, -5.0, -5.0, 0.0];
        let all = LogProbabilities::new(&values, 3, 3).unwrap();
        let first = LogProbabilities::new(&values[..6], 2, 3).unwrap();
        let second = LogProbabilities::new(&values[6..], 1, 3).unwrap();
        let mut one_shot = CTCDecoder::new(DecoderConfig::default()).unwrap();
        let expected = one_shot
            .prefix_beam_search(
                all,
                PrefixBeamSearchOptions {
                    finalize: true,
                    ..PrefixBeamSearchOptions::default()
                },
            )
            .unwrap();
        let mut streaming = CTCDecoder::new(DecoderConfig::default()).unwrap();
        streaming
            .prefix_beam_search(first, PrefixBeamSearchOptions::default())
            .unwrap();
        let actual = streaming
            .prefix_beam_search(
                second,
                PrefixBeamSearchOptions {
                    finalize: true,
                    ..PrefixBeamSearchOptions::default()
                },
            )
            .unwrap();
        assert_eq!(actual, expected);
    }

    #[test]
    fn context_completion_changes_ranking() {
        let values = [-4.0, -0.1, -2.0, -4.0, -2.0, -0.2];
        let input = LogProbabilities::new(&values, 2, 3).unwrap();
        let mut decoder = CTCDecoder::new(DecoderConfig {
            context_token_ids: vec![vec![1, 2]],
            context_policy: Some(ContextPolicy {
                completion_bonus: 5.0,
                ..ContextPolicy::conservative()
            }),
            ..DecoderConfig::default()
        })
        .unwrap();
        let result = decoder
            .prefix_beam_search(
                input,
                PrefixBeamSearchOptions {
                    return_scores: true,
                    ..PrefixBeamSearchOptions::default()
                },
            )
            .unwrap();
        assert_eq!(result.tokens[0], vec![1, 2]);
        assert_eq!(result.scores.unwrap()[0].contextual_bias, 5.0);
    }
}
